### PR TITLE
Add metric descriptions

### PR DESCRIPTION
**Pull request by ufukyilmaz**
This PR add metric description appendix to the docs. For now, I've added all of the metric to the table but there could be still some missing pieces (I believe there is no). Also, I do not understand some metrics enough to document that's why I left their descriptions empty for now.
 
Related issues: 
https://github.com/hazelcast/hazelcast/issues/17880
https://github.com/hazelcast/imdg-docs/issues/131

### CODE REVIEW

File: docs/modules/ROOT/pages/metrics.adoc [hunk 1]
```diff
@@ -0,0 +1,842 @@
...
+|the number of destroyed proxies for a given service.
+
+|`proxy.proxyCount`
+|count
+|the number of active proxies for a given service. e.g., the number of all proxies for the IMap.
+
+
+|`raft.destroyedGroupIds`
+|count
+|
```
Comment (blazember) [inline docs/modules/ROOT/pages/metrics.adoc:652]: Do you need help with these?

File: docs/modules/ROOT/pages/metrics.adoc [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+|ms
+|the uptime of the JVM
+
+|`runtime.usedMemory`
+|bytes
+|an approximation to the total amount of memory currently used
+
+|`tcp.acceptor.eventCount`
+|count
+|
```
Comment (blazember) [inline docs/modules/ROOT/pages/metrics.adoc:704]: the total number of the connections accepted by TcpServerAcceptor

File: docs/modules/ROOT/pages/metrics.adoc [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+|count
+|the number of accepted socket channels
+
+|`tcp.connection.activeCount`
+|count
+|the number of active connections
+
+|`tcp.connection.clientCount`
+|count
+|
```
Comment (blazember) [inline docs/modules/ROOT/pages/metrics.adoc:744]: the number of the active client connections

File: docs/modules/ROOT/pages/metrics.adoc [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+|count
+|the number of active connection listeners
+
+|`tcp.connection.count`
+|count
+|the number of `TcpServerConnection`
+
+|`tcp.connection.inProgressCount`
+|count
+|
```
Comment (blazember) [inline docs/modules/ROOT/pages/metrics.adoc:760]: The number of connections establishments in progress.

File: docs/modules/ROOT/pages/metrics.adoc [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+
+|tcp.connection.openedCount
+|count
+|the number of opened connections
+
+|tcp.connection.textCount
+|count
+|
```
Comment (blazember) [inline docs/modules/ROOT/pages/metrics.adoc:768]: The number of connections used by text-based protocols (REST, Memcache).

**Review by blazember — COMMENTED** [review]
Thanks for picking this up and documenting the metrics :tada: 
Added some comments to consider and some questions. Let me know if you need help in figuring out the meaning of some metrics with cryptic names :smile:

**Comment by JakeSCahill** [comment]
Nice, this is great.

I haven't had a chance to read through this yet, but just adding a note to myself or @Serdaro to remember to add this to the navigation.

The navigation file is here: https://github.com/hazelcast/imdg-docs/blob/master/docs/modules/ROOT/nav.adoc

File: docs/modules/ROOT/pages/metrics.adoc (hunk 1 above)
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc:960]: I need help with them. I can only convert the metric name to the description but this won't be explanatory.

**Comment by Serdaro** [comment]
@ufukyilmaz , @mmedenjak , I am guessing it is ready for review now?

**Comment by JakeSCahill** [comment]
@ufukyilmaz could you pull the latest changes to `master` into your branch? We now have a staging site that builds a preview of any changes. To get the build to work, you need the latest changes.

**Comment by ufukyilmaz** [comment]
For now, there are some missing metric documentation:
- I couldn't document raft metrics
- I couldn't collect some of the client metrics (so I couldn't be sure about the exact names of them and there may be another missing metrics): https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/metrics/MetricDescriptorConstants.java#L66-L73
- And also I missed documenting nio pipeline metrics (I will add them soon)

File: docs/modules/ROOT/pages/metrics.adoc [hunk 6]
```diff
@@ -0,0 +1,842 @@
...
+|count
+|the number of partition operation handler threads for given member
+
+|`operation.priorityQueueSize`
+|count
+|the number of priority operations pending (priority partition ops. + priority generic ops.)
+
+|`operation.queueSize`
+|count
+|
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc]: Added them

**Comment by mmedenjak** [comment]
> ufukyilmaz , mmedenjak , I am guessing it is ready for review now?

Just going through my email backlog :) Yes, I believe it's ready :)

**Comment by mmedenjak** [comment]
I just saw that even the CP metrics were added. Really nice work @ufukyilmaz , you're awesome 🏅

File: docs/modules/ROOT/pages/metrics.adoc [hunk 7]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|the connection close time for this connection
+
+|`cluster.connection.connectionId`
+|count
+|the connection id for this client connection
+
+|`cluster.connection.eventHandlerCount`
+|count
+|the number of event handler for this client connection
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:127]: Number of the event handlers ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 8]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cluster.heartbeat.lastHeartbeat`
+|ms
+|the last timestamp of sending heartbeats to the other known cluster members
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:131]: Last timestamp of the heartbeats sent to the ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 9]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cluster.size`
+|count
+|the size of the cluster
+
+|`cp.atomiclong.value`
+|count
+|the value of this AtomicLong
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:139]: Maybe making AtomicLong as monospaced as in the other primitives mentioned below, i.e., `IAtomicLong`

File: docs/modules/ROOT/pages/metrics.adoc [hunk 10]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.countdownlatch.count`
+|count
+|the initial count of the `ICountDownLatch`
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:143]: ... count of `ICountDown...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 11]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.countdownlatch.remaining`
+|count
+|the remaining number of expected count-downs
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:147]: ... expected countdowns

File: docs/modules/ROOT/pages/metrics.adoc [hunk 12]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.countdownlatch.round`
+|count
+|the round number of the `ICountDownLatch`, each time `ICountDownLatch` is initialized with a new count after it downs to zero, a new round begins
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:151]: Round number of `ICountDownLoatch`; each time ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 13]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.lock.acquireLimit`
+|count
+|the maximum number of reentrant acquires of this `FencedLock`
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:155]: ... reentrant acquirements by this ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 14]
```diff
@@ -0,0 +1,1678 @@
...
+|`cp.lock.ownerSessionId`
+|count
+|the sessionId of the `FencedLock` owner
+
+|`cp.semaphore.available`
+|count
+|the number of the remaining available permits
+
+|`cp.semaphore.initialized`
+|count
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:174]: boolean

File: docs/modules/ROOT/pages/metrics.adoc [hunk 15]
```diff
@@ -0,0 +1,1678 @@
...
+|a boolean state, which shows whether semaphore is initialized with a value or not. In exposed metrics it will be 0 when semaphore is not initialized, a positive value otherwise
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:175]: Specifies whether .... or not; in exposed metrics, it shows 0 when ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 16]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.session.creationTime`
+|ms
+|the creation time of this session
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:179]: ... this CP session

File: docs/modules/ROOT/pages/metrics.adoc [hunk 17]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.session.endpoint`
+|
+|the address of the endpoint which session belongs to
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:183]: ... which the CP session ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 18]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.session.endpointType`
+|
+|the type of the endpoint; either SERVER or CLIENT
+
+|`cp.session.expirationTime`
+|ms
+|the expiration time of the session
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:191]: ... of the CP session

File: docs/modules/ROOT/pages/metrics.adoc [hunk 19]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`cp.session.version`
+|count
+|a boolean state, which shows whether semaphore is initialized with a value or not. In exposed metrics it will be 0 when semaphore is not initialized, a positive value otherwise
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:195]: Version of the CP session (maybe a clear explanation of what "version" means here).

File: docs/modules/ROOT/pages/metrics.adoc [hunk 20]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|the number of subscribed listeners for the specified service
+
+|`event.publicationCount`
+|count
+|the number of published events for the specified service
+
+|`event.queueCapacity`
+|count
+|the queue capacity of the executor processing the events. This capacity is shared for all events
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:215]: ... events (this capacity .....)

File: docs/modules/ROOT/pages/metrics.adoc [hunk 21]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`event.rejectedCount`
+|count
+|the number of rejected events. If the event is not accepted to the executor in `hazelcast.event.queue.timeout.millis`(see xref:system-properties.adoc[]), it will be rejected and not processed
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:219]: Number of the rejected events; if the event is not accepted by the executor in `hazelcast .....

(It'd be good to place an anchor on the related system property to forward the reader directly to the description of it in the System Properties appendix, I'll take care of it).

File: docs/modules/ROOT/pages/metrics.adoc [hunk 22]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`event.syncDeliveryFailureCount`
+|count
+|the number of failures of sync event delivery
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:223]: I couldn't get this. The topic is like the delivery of synced events?

File: docs/modules/ROOT/pages/metrics.adoc [hunk 23]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|the number of pending operations on the executor service
+
+|`executor.started`
+|count
+|the number of started operations on the executor service
+
+|`executor.totalExecutionTime`
+|ms
+|the total execution time of operations finished
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:255]: ... time of the finished operations

**Review by Serdaro — COMMENTED** [review]
My comments - Part 1

File: docs/modules/ROOT/pages/metrics.adoc [hunk 24]
```diff
@@ -0,0 +1,1470 @@
...
+|count
+|the number of gets on the cache
+
+|`cache.cacheHits`
+|count
+|the number of hits (successful get operations) on the cache
+
+|`cache.cacheHitPercentage`
+|percent
+|the percentage of hits (successful get operations) on the cache
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc:39]: yes, exactly

**Comment by emre-aydin** [comment]
We were missing this for a long time. Great job @ufukyilmaz, an invaluable contribution! Could you also please document the tags for each metric, similar to how it's done for Jet (https://jet-start.sh/docs/operations/monitoring#cluster-wide-metrics - no need for grouping by tags though)?

As discussed in our call, I will wait until they are added before reviewing. @blazember I suggest you do the same. @ufukyilmaz please make sure to ask for a review again when you're done with your changes so that we're notified.

File: docs/modules/ROOT/pages/metrics.adoc (hunk 24 above)
[the remark this answers is not in the document]
Comment (Serdaro) [reply docs/modules/ROOT/pages/metrics.adoc]: Then let's mention it like that

File: docs/modules/ROOT/pages/metrics.adoc (hunk 14 above)
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc:174]: I updated its description

File: docs/modules/ROOT/pages/metrics.adoc [hunk 25]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Maximum number of open file descriptors (only for UNIX platforms)
+
+|`os.openFileDescriptorCount`
+|count
+|Number of open file descriptors (only for UNIX platforms)
+
+|`os.processCpuLoad`
+|percent
+|Recent cpu usage for the JVM process; a negative value if not available
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:980]: Recent CPU ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 26]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`os.processCpuTime`
+|ms
+|CPU time used by the process on which the JVM is running
+
+|`os.systemCpuLoad`
+|percent
+|Recent cpu usage for the whole system; a negative value if not available
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:988]: Recent CPU ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 27]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Number of owned items in this member
+
+|`queue.total`
+|count
+|Total number of operations (numberOfOffers + numberOfPolls + numberOfOtherOperations)
+
+|`raft.destroyedGroupIds`
+|count
+|Destroyed raft node group ids
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1180]: ... group IDs

File: docs/modules/ROOT/pages/metrics.adoc [hunk 28]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Number of missing CP members
+
+|`raft.nodes`
+|count
+|Number of local Raft nodes
+
+|`raft.terminatedRaftNodeGroupIds`
+|count
+|terminated raft node group ids
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1236]: Terminated ... group IDs

File: docs/modules/ROOT/pages/metrics.adoc [hunk 29]
```diff
@@ -0,0 +1,1678 @@
...
+
+|`replicatedMap.creationTime`
+|ms
+|Creation time of this replicated map on this member.
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1240]: ... on this member

File: docs/modules/ROOT/pages/metrics.adoc [hunk 30]
```diff
@@ -0,0 +1,1678 @@
...
+|ms
+|Last update time of the locally owned items
+
+|`tcp.acceptor.eventCount`
+|count
+|Total number of the connections accepted by TcpServerAcceptor
+
+|`tcp.acceptor.exceptionCount`
+|count
+|Number of thrown exception on this TcpServerAcceptor
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1352]: ... thrown exceptions ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 31]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Number of active connection listeners
+
+|`tcp.connection.count`
+|count
+|Number of `TcpServerConnection`
+
+|`tcp.connection.inProgressCount`
+|count
+|Number of connections establishments in progress
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1404]: ... connection establishments ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 32]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Total number of the connections accepted by TcpServerAcceptor
+
+|`tcp.inputThread/outputThread.framesTransceived`
+|count
+|Number of transceived frames on this NioThread
+
+|`tcp.inputThread/outputThread.idleTimeMillis`
+|ms
+|Idle time that indicates how long since the last read/write
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1508]: ... indicates the duration since the last ...

File: docs/modules/ROOT/pages/metrics.adoc [hunk 33]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Number of transceived priority frames
+
+|`tcp.inputThread/outputThread.processCount`
+|count
+|Number of processed `NioPipeline`s on this NioThread
+
+|`tcp.inputThread/outputThread.selectorIOExceptionCount`
+|count
+|Number of times that io exceptions are thrown during selection
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1524]: ... that I/O exceptions..

(just being consistent with the other I/O mentions in the documentation)

**Review by Serdaro — COMMENTED** [review]
My comments.. Part II

File: docs/modules/ROOT/pages/metrics.adoc [hunk 34]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Total number of minor GCs that have occurred
+
+|`gc.minorTime`
+|ms
+|Accumulated elapsed time in minor GCs
+
+|`gc.unknownCount`
+|count
+|Number of unknown GCs that cannot be determined to be as a minor or major (this is usually due to the lack of support of the used garbage collector)
```
Comment (JakeSCahill) [inline docs/modules/ROOT/pages/metrics.adoc:324]:
```suggestion
|Number of unknown GCs that cannot be determined as minor or major (this is usually due to the lack of support of the used garbage collector)
```

File: docs/modules/ROOT/pages/metrics.adoc [hunk 35]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Total number of index hits (the value of this metric may be greater than the `map.index.queryCount` since a single query may hit the same index more than once)
+
+|`map.index.insertCount`
+|count
+|Number of insert operations performed on the index
+
+|`map.index.memoryCost`
+|bytes
+|Local memory cost of the index (for on-heap indexes (OBJECT and BINARY storages), the returned value is just a best-effort approximation and doesn't indicate a precise on-heap memory usage of the index)
```
Comment (JakeSCahill) [inline docs/modules/ROOT/pages/metrics.adoc:524]:
```suggestion
|Local memory cost of the index (for on-heap indexes in OBJECT or BINARY formats, the returned value is just a best-effort approximation and doesn't indicate a precise on-heap memory usage of the index)
```

**Review by JakeSCahill — COMMENTED** [review]
general comments and suggestions to add clarity

File: docs/modules/ROOT/pages/metrics.adoc [hunk 36]
```diff
@@ -0,0 +1,1678 @@
...
+|ms
+|Member's local clock timestamp
+
+|`cluster.clock.maxClusterTimeDiff`
+|ms
+|Maximum observed cluster time difference
+
+|`cluster.connection.closedTime`
+|count
+|Connection close time for this connection
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc:119]: for some metrics, there are also a metric description tags that shows what this metric belongs to, for this case, these 4 tags are specified what this connection is:
- endpoint=[10.212.134.152]:5701
- client=bdae5ad8-4be6-4217-87e7-15dd04a237dd
- clientname=hz.client_1
- timestamp=1616759245227

but I did not document them in this first iteration

File: docs/modules/ROOT/pages/metrics.adoc [hunk 37]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Round number of the `ICountDownLatch`; each time `ICountDownLatch` is initialized with a new count after it downs to zero, a new round begins
+
+|`cp.lock.acquireLimit`
+|count
+|Maximum number of reentrant acquires of this `FencedLock`
+
+|`cp.lock.lockCount`
+|count
+|Number that shows how many times this `FencedLock` acquired
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc]: Addressed in 5c51e3b

File: docs/modules/ROOT/pages/metrics.adoc [hunk 38]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Number of event handlers for this client connection
+
+|`cluster.heartbeat.lastHeartbeat`
+|ms
+|Last timestamp of sending heartbeats to the other known cluster members
+
+|`cluster.size`
+|count
+|Size of the cluster
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc]: Addressed in 5c51e3b

File: docs/modules/ROOT/pages/metrics.adoc [hunk 39]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Number of gets on the cache
+
+|`cache.cacheHits`
+|count
+|Number of hits (successful get operations) on the cache
+
+|`cache.cacheHitPercentage`
+|percent
+|Percentage of hits (successful get operations) out of all get operations on the cache
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc]: addressed in b0330f3

File: docs/modules/ROOT/pages/metrics.adoc [hunk 40]
```diff
@@ -0,0 +1,1678 @@
...
+|count
+|Maximum number of threads in the executor's thread pool
+
+|`executor.internal.poolSize`
+|count
+|Number of threads in the executor's thread pool
+
+|`executor.internal.queueSize`
+|count
+|Number
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc]: added the description of it (number of pending task in this executor's task queue)

**Review by Serdaro — APPROVED** [review]

**Comment by Serdaro** [comment]
@blazember , @emre-aydin , Could you please take a look after the recent changes?

**Review by emre-aydin — APPROVED** [review]
Great work @ufukyilmaz, hugely important!

**Comment by emre-aydin** [comment]
What are we missing for merging this PR?

File: docs/modules/ROOT/pages/metrics.adoc [hunk 41]
```diff
@@ -0,0 +1,1681 @@
+= Metrics
```
Comment (Serdaro) [inline docs/modules/ROOT/pages/metrics.adoc:1]: Can we change the title to "List of Hazelcast Metrics"?

Comment (ufukyilmaz) [reply docs/modules/ROOT/pages/metrics.adoc]: Made the requested change

**Comment by mmedenjak** [comment]
Thank you all for your work on this PR and especially thank you to @ufukyilmaz for driving it forward :)

**Pull request merged**